qcodes_contrib_drivers/drivers/Vaunix/LDA.py

```diff
@@ -0,0 +1,299 @@
+
+
+class Vaunix_LDA(Instrument):
+    dll_path = None
+
+    def __init__(self, name: str,
+                 serial_number: int,
+                 dll_path: Optional[str] = None,
+                 channel_names: Optional[Dict[int, str]] = None,
+                 test_mode: bool = False,
+                 **kwargs):
+        r"""
+        QCoDeS Instrument for Vaunix LDA digital attenuators.
+
+        Args:
+            name: Qcodes name for this instrument
+            serial_number: Serial number of the instrument, used to identify
+                it.
+            dll_path: Look for the LDA DLLs in this directory. Sets the dll
+                path as class attribute that is used for future instances for
+                which ``dll_path`` is not given.
+            channel_names: Optionally assign these names to the channels.
+            test_mode: If True, simulates communication with an LDA-102
+                (serial:55102). Does not communicate with physical devices. For
+                testing purposes.
+        """
+
+        super().__init__(name=name, **kwargs)
+        self.serial_number = serial_number
+        self.reference = None
+
+        if channel_names is None:
+            channel_names = {}
+
+        self.dll = self._get_dll(dll_path)
+        self.dll.fnLDA_SetTestMode(test_mode)  # Test API without communication
+
+        # Find all Vaunix devices, init the one with matching serial number.
+        num_devices = self.dll.fnLDA_GetNumDevices()
+        device_IDs = ctypes.c_int * num_devices
+        device_refs = device_IDs()
+        self.dll.fnLDA_GetDevInfo(device_refs)
+        devices = {self.dll.fnLDA_GetSerialNumber(ref): ref
+                   for ref in device_refs}
+        self.reference = devices.get(self.serial_number, "not found")
+        if self.reference == "not found":
+            raise ValueError(f"LDA with serial number {self.serial_number}"
+                             f" was not found in the system. Found: {devices}")
+
+        self.dll.fnLDA_InitDevice(self.reference)
+
+        num_channels = self.dll.fnLDA_GetNumChannels(self.reference)
+        if num_channels == 1:
+            # don't add Channel objects, add parameters directly instead
+            _add_lda_parameters(self)
+        else:
+            for i in range(1, num_channels + 1):
+                name = channel_names.get(i, f"ch{i}")
+                ch = LdaChannel(parent=self, channel_number=i, name=name)
+                self.add_submodule(name, ch)
+
+        self.connect_message()
+
+    def _get_dll(self, dll_path: str = None) -> ctypes.CDLL:
+        r"""
+        Load correct DLL from ``dll_path`` based on bitness of the operating
+        system.
+
+        Args:
+            dll_path: path to the directory that contains the Vaunix LDA DLL.
+                By default, use class attribute ``Vaunix_LDA.dll_path``.
+        """
+        Vaunix_LDA.dll_path = dll_path or Vaunix_LDA.dll_path
```

Comment:
in this case, it might not be smart to set the class attribute from an instance. my idea was that the user either specifies `Vaunix_LDA.dll_path` before initializing OR passes `dll_path` to `__init__`. But it's up to you, what you think is most appropriate in this case.

Reply:
Fair point, updated in 46a9c13.